Add the ability to order operations performed by the batcher

Calastrophe opened the pull request:
As it stands, there isn't a way to change the ordering in which memory operations are performed by the batcher when committed to memory.

This pull requests seeks to remedy this by keeping the prior default behavior of the batcher, but allows for people to change the behavior by explicitly setting the ordering (on the fly) before subsequent calls to `commit_rw()`.

ko1N commented:
Another idea that just came into my mind is that we could mimic the actual ordering of how end users invoke the batcher. We could keep only one list instead of a read + write list (maybe by mapping it via an enum) - this way we could keep the ordering the user providers as well. Maybe this could be a seperate flag as well.

Calastrophe commented:
I've added the requested changes and added a bit more to the documentation to reflect the changes overall.

Additionally, I tried a couple different iterations of trying to keep the ordering of the batcher without incurring some sort of overhead. In the future, if someone wanted to add it I think keeping the read/write list separate and then tracking order using an enum with an index into each list.

Something like,

```rs
enum Operation {
    Read { index: usize },
    Write { index: usize },
}
```

then, you could just have this stored in its own array keeping the order and if they've selected the sequential ordering, you can group operations and then batch the groups as best as you can.

However, if they've constantly flip-flopped between reading and writing all that work is for nil.
